File: SDE_sheet/Arrays/12_count_inversions/solution.py
def number_of_inversions(a):
    return merge_sort(a, 0, len(a) - 1)

def merge_sort(arr, low, high):
    cnt = 0

    if low >= high:
        return cnt

    mid = low + (high - low) // 2

    cnt += merge_sort(arr, low, mid)
    cnt += merge_sort(arr, mid + 1, high)
    cnt += merge(arr, low, mid, high)

    return cnt

def merge(arr, low, mid, high):
    temp = []
    left = low
    right = mid + 1
    cnt = 0

    while left <= mid and right <= high:
        if arr[left] <= arr[right]:
            temp.append(arr[left])
            left += 1
        else:
            temp.append(arr[right])
            cnt += mid - left + 1
            right += 1

    while left <= mid:
        temp.append(arr[left])
        left += 1

    while right <= high:
        temp.append(arr[right])
        right += 1

    for i in range(low, high + 1):
        arr[i] = temp[i - low]

    return cnt

File: SDE_sheet/Arrays/12_count_inversions/solution.java
public class InversionCount {
    public int numberOfInversions(int[] a) {
        return mergeSort(a, 0, a.length - 1);
    }

    private int mergeSort(int[] arr, int low, int high) {
        int cnt = 0;

        if (low >= high) {
            return cnt;
        }

        int mid = low + (high - low) / 2;

        cnt += mergeSort(arr, low, mid);
        cnt += mergeSort(arr, mid + 1, high);
        cnt += merge(arr, low, mid, high);

        return cnt;
    }

    private int merge(int[] arr, int low, int mid, int high) {
        int[] temp = new int[high - low + 1];
        int left = low, right = mid + 1, cnt = 0, k = 0;

        while (left <= mid && right <= high) {
            if (arr[left] <= arr[right]) {
                temp[k++] = arr[left++];
            } else {
                temp[k++] = arr[right++];
                cnt += mid - left + 1;
            }
        }

        while (left <= mid) {
            temp[k++] = arr[left++];
        }

        while (right <= high) {
            temp[k++] = arr[right++];
        }

        for (int i = 0; i < temp.length; i++) {
            arr[low + i] = temp[i];
        }

        return cnt;
    }
}

